Implement Camera

LunaTheFoxgirl opened the pull request:
- [x] Depends on #20.
- Closes #7.

This PR aims to add Camera support to Inox2D, still need to figure out how we pass in the viewport

TODO:
- [x] Implement Camera struct
- [x] Implement viewport update code
- [x] Implement shader side viewport

LunaTheFoxgirl committed:
Implement camera struct

LunaTheFoxgirl committed:
Not using projection cache anyways

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,49 @@
+    /// Gets the center offset of the viewport
+    pub fn get_center_offset(&mut self, viewport: Vec2) -> Vec2 {
+        let real_size = self.get_real_size(viewport);
+        Vec2 { x: (real_size.x/2.0), y: (real_size.y/2.0) }
```
```suggestion
        real_size / 2.0
```

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,49 @@
+        let real_size = self.get_real_size(viewport);
+
+        // Faster to reuse real_size, so do that instead of calling get_center_offset
+        let origin = Vec2{ x: real_size.x/2.0, y: real_size.y/2.0 };
```
```suggestion
        let origin = real_size / 2.0;
```

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,49 @@
+
+        // Faster to reuse real_size, so do that instead of calling get_center_offset
+        let origin = Vec2{ x: real_size.x/2.0, y: real_size.y/2.0 };
+        let pos = Vec3{ x: self.position.x, y: self.position.y, z: -(u16::MAX as f32)/2.0};
```
```suggestion
        let pos = self.position.extend(-(u16::MAX as f32) / 2.0);
```

Maybe even potentially this?
```suggestion
        let pos = self.position.extend(i16::MIN as f32);
```

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,49 @@
+        let pos = Vec3{ x: self.position.x, y: self.position.y, z: -(u16::MAX as f32)/2.0};
+
+        // Return camera ortho matrix
+        Mat4::orthographic_lh(0.0, real_size.x, real_size.y, 0.0, 0.0, u16::MAX as f32) * 
```
We can omit the zero decimals in Rust :)
```suggestion
        Mat4::orthographic_lh(0., real_size.x, real_size.y, 0., 0., u16::MAX as f32) * 
```

LunaTheFoxgirl replied:
I find 0.0 to be easier to read than 2 punctuation marks right next to each other

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,49 @@
+
+        // Return camera ortho matrix
+        Mat4::orthographic_lh(0.0, real_size.x, real_size.y, 0.0, 0.0, u16::MAX as f32) * 
+            Mat4::from_translation(Vec3{x: origin.x, y: origin.y, z: 0.0}) *
```
```suggestion
            Mat4::from_translation(origin.extend(0.)) *
```

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,49 @@
+impl Default for Camera {
+    fn default() -> Self {
+        Self {
+            position: Vec2{ x: 0.0, y: 0.0 },
```
```suggestion
            position: Vec2::ZERO,
```

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,49 @@
+    /// Gets the real size of the viewport
+    pub fn get_real_size(&mut self, viewport: Vec2) -> Vec2 {
+        Vec2 {
+            x: viewport.x/self.scale.x, 
+            y: viewport.y/self.scale.y,
```
I recommend you format the code with `cargo fmt`, or CTRL+SHIFT+I since you use VSCode.
```suggestion
            x: viewport.x / self.scale.x, 
            y: viewport.y / self.scale.y,
```

LunaTheFoxgirl committed:
Improve code style

LunaTheFoxgirl committed:
Run cargofmt on camera code

Speykious commented on src/math/camera.rs:
```diff
@@ -0,0 +1,48 @@
+
+    /// Gets the center offset of the viewport
+    pub fn get_center_offset(&mut self, viewport: Vec2) -> Vec2 {
+        let real_size = self.get_real_size(viewport);
+        real_size / 2.0
```
Just realized
```suggestion
        self.get_real_size(viewport) / 2.0
```

Speykious commented:
I'll see what I can do to put the mvp into the shaders.
That means I also need to support resizing the composite frameuffer (which link mauve has already done, but he doesn't want to use an mvp).

Speykious committed:
Prefer no `get_` prefix for getters

Speykious committed:
Inline variable in `center_offset`

Speykious committed:
Support viewport resizing

Speykious committed:
Merge branch 'main' into camera

Speykious committed:
Don't put uniform location into an option

Speykious committed:
Make Camera derive Clone

Speykious committed:
Unnecessary mut

Speykious committed:
Implement MVP in renderer

This adds a camera to the renderer and
modifies the simple shaders in a way to integrate
an MVP without modifying too many things.

Speykious committed:
Add zoom in `render_inp` example

Speykious committed:
Add cursor world movement

Speykious committed:
Cache camera (it was worth it)

Speykious reviewed (approved):
Should look good now
